### ISSUE TITLE
PyPy virtualenv headers not available

### ISSUE THREAD

**Issue by Czaki** [joerick/cibuildwheel#501]
I meet the next bug in PyPy creating virtualenv. File `Python.h` is not available form virtualenv, which breaks compilation of packages which does not provide PyPy wheels:

https://dev.azure.com/bokota/imagecodecs/_build/results?buildId=565&view=logs&j=9ea3a1cf-027d-5faa-88b4-e0b2b34fabd3&t=77745ba4-e9ed-58b1-0e6d-5824980d22ee&l=249

https://dev.azure.com/bokota/imagecodecs/_build/results?buildId=565&view=logs&j=c436af4d-dd17-5b6b-70ac-85e5b57c9f0c&t=bc0cc3b0-6386-533a-de3c-d01e46ec05cd&l=279

Did You remember who should be called for this? Would we like to add a test that simulates this? I do not want to open heptapod account (which I will forget in short time).

**Comment by YannickJadoul** [comment]
> I do not want to open heptapod account (which I will forget in short time).

Seems you can login by linking to your GitHub account. But if you have a minimal reproducer, I can report it, if you prefer.

**Comment by Czaki** [comment]
```
python -m pip install scikit-build
python -m pip install blosc==1.9.2
```
Works without virtualenv and fail in virtualenv with 

```
 ../../../blosc/blosc_extension.c:12:10: fatal error: Python.h: No such file or directory
     12 | #include "Python.h"
```

Happens on pypy3.6 and pypy3.7.

**Comment by henryiii** [comment]
This seems vaguely familiar... Don't remember where I've seen this before. Pretty sure it had something to do with a PyPy venv there too.

**Comment by henryiii** [comment]
You should poke the blosc authors to add a pyproject.toml to install Scikit-Build instead of having to do it manually.

**Comment by henryiii** [comment]
Scikit-build does get the include directory for PyPy and pass it on, but it might not do that properly for PyPy? It happens to work outside the venv because the includes are in a common spot. Have you compared to CPython's venv and includes location?

**Comment by Czaki** [comment]
I test it. I have no problem with cPython virtualenv. I also install pypy using pyenv, so it is outside the standard location.

**Comment by henryiii** [comment]
What I meant was:

* where is Python.h for CPython venv? In venv or outside?
* If inside, is it missing for PyPy?

**Comment by henryiii** [comment]
Code is here: https://github.com/scikit-build/scikit-build/blob/c0741f9a1de9dd97e0a1e6f0b3b02fcd76221418/skbuild/cmaker.py#L262

There should be a Python in `sysconfig.get_config_var('INCLUDEPY')`, inside the env it should report the same thing as outside. At least it should, I can't make the pypy3 venv:

```bash
pypy3 -m venv venvpypy
Error: Command '['/Users/henryiii/venvpypy/bin/pypy3', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1.
```

If you can make one and try it, perhaps?

**Comment by henryiii** [comment]
PS: I'm on Big Sur, latest homebrew, Intel, PyPy3 3.7.2. So it's quite likely that I've got an old pip from ensurepip, and you have to have 20.3 to get wheels on Big Sur (even Intel), and with a broken environment (incorrect Python include dir perhaps), it can't update from source. @mattip would probably be the person to talk to.

**Comment by mattip** [comment]
@henryiii I am not sure what your BigSur machine has to do with the azure build failure.

The first failing compilation command is [here](https://dev.azure.com/bokota/imagecodecs/_build/results?buildId=565&view=logs&j=9ea3a1cf-027d-5faa-88b4-e0b2b34fabd3&t=77745ba4-e9ed-58b1-0e6d-5824980d22ee&l=442), it is
```
/opt/rh/devtoolset-8/root/usr/bin/cc -Dblosc_extension_EXPORTS \
    -I/tmp/tmp.wC6BaqoJgL/venv/include 
    -I../../../c-blosc/blosc \
    -I../../../c-blosc/internal-complibs/lz4-1.9.2 \
    -I../../../c-blosc/internal-complibs/zlib-1.2.8 \
    -I../../../c-blosc/internal-complibs/zstd-1.4.5 \
    -I../../../c-blosc/internal-complibs/zstd-1.4.5/common \
    -O3 -DNDEBUG -fPIC -MD -MT \
    CMakeFiles/blosc_extension.dir/blosc/blosc_extension.c.o \
    -MF CMakeFiles/blosc_extension.dir/blosc/blosc_extension.c.o.d \
    -o CMakeFiles/blosc_extension.dir/blosc/blosc_extension.c.o \
    -c ../../../blosc/blosc_extension.c
```
Note the include path `-I/tmp/tmp.wC6BaqoJgL/venv/include ` which is inside the virtualenv. When I create a pypy virtualenv, for some reason the include path inside it has **all** the header files from the original non-virtualenv **except** `Python.h`. Is this a virtualenv bug?

**Comment by mattip** [comment]
My bad, I wasn't looking carefully. The `Python.h` file is there. So why is the compilation failing?

**Comment by mattip** [comment]
Got it. virtualenv-20.0.31 does copy the include files, virtualenv-20.2.2 does not.

**Comment by mattip** [comment]
... and on CPython, `python -c"import sysconfig; print(sysconfig.get_config_var('INCLUDEPY'))"` will point to the non-virtualenv location.

**Comment by henryiii** [comment]
> I am not sure what your BigSur machine has to do with the azure build failure.

I just meant that I wasn’t able to easily run the test I was asking you to run, as it was broken earlier in the chain for me.

> virtualenv-20.0.31

Great, do we need to do a pinning update? Or is this baked into PyPy? On a phone for now, so can’t as easily check.

**Comment by henryiii** [comment]
Oh, it’s the other direction, we we need a temporary downgrade + bug report?

It seems like it should not copy the files and instead point to the proper place, like CPython?

**Comment by mattip** [comment]
This is a PyPy bug. If you could pin virtualenv that would be great, at least until a new version of PyPy is released with the fix from [the issue](https://foss.heptapod.net/pypy/pypy/-/issues/3364) I filed there.

**Comment by mattip** [comment]
Thanks for the report!

**Comment by henryiii** [comment]
Can we do a PyPy specific pin, or will our update infrastructure blow up? I assume this has come up before and can be done?

**Comment by mattip** [comment]
The build log is from scikit-build. If they want to support PyPy, they should be [pinning virtualenv to 20.0.31](https://github.com/scikit-build/scikit-build/blob/master/azure-pipelines.yml#L83)

**Comment by henryiii** [comment]
@jcfr

Though we pin it unless asked not to so we could do it too until we update. I’d like to see Scikit-build, CMake, and ninja use CI build wheel and would be up for helping with it. ;)

**Comment by jcfr** [comment]
@henryiii thanks for offering to help 🙏give me few days to assess the situation and I will get back to you.

**Comment by Czaki** [comment]
> The build log is from scikit-build. If they want to support PyPy, they should be [pinning virtualenv to 20.0.31](https://github.com/scikit-build/scikit-build/blob/master/azure-pipelines.yml?rgh-link-date=2020-12-24T21%3A34%3A22Z#L83)

In my case, the cibuildwheel creates virtualenv, not scikit-build.

**Comment by henryiii** [comment]
Looks like we just use pip compile, and that only uses the current environment, that is, I don't think it will propagate environment markers. Where it makes this:

```
virtualenv==20.2.2        # via -r cibuildwheel/resources/constraints.in
```

Instead we want this:

```
virtualenv==20.2.2 ; platform_python_implementation!="PyPy"
virtualenv==20.0.31 ; platform_python_implementation=="PyPy"
```

How long would we be likely to have to maintain something like that? Very short term, it could be by hand, longer term, we'd need some post processing, perhaps?

**Comment by joerick** [comment]
It's amazing how you're so knowledgeable about these things @henryiii!!  I did not know about it but `platform_python_implementation`in the requirement would be perfect for this. Does pip-compile understand such a flag? If so we could introduce a constraints-pypy36.txt that's generated from our tooling. If not, this can just be a manual change we make for now, or (assuming there are no downsides) we could pin CPython to the older version too.

**Comment by henryiii** [comment]
The table is here: https://www.python.org/dev/peps/pep-0508/#environment-markers

Unfortunately, I’ve needed this to write messes like this: https://github.com/pybind/pybind11/blob/master/tests/requirements.txt

pip-compile evaluates all environment markers AFAICT from the docs (and keeping them would be hard, I think - the point is to compile based on the current environment).

Pinning an old version might cause issues on Big Sur, as it has an internal pip copy, I believe, that it installs in fresh environments.

**Comment by henryiii** [comment]
If we do want to split files based on PyPy vs CPython, yes, pip-compile would work fine.

**Comment by Czaki** [comment]
I'm not sure if adding the next 3 files (pypy27, pypy36, pypy37) is a good idea as a final solution. It looks like a nice workaround, for this problem, but I prefer to treat it as a temporary workaround.

**Comment by henryiii** [comment]
This will break at least once a week when we do our scheduled update, plus once per release. How temporary is an important question.

**Comment by mattip** [comment]
The need will go away with the next release of PyPy, probably in a month or two.

**Comment by YannickJadoul** [comment]
> This will break at least once a week when we do our scheduled update, plus once per release. How temporary is an important question.

Can't we just apply a patch at the end of the script, or so?

**Comment by mattip** [comment]
The patch would look like [the pure python fix](https://foss.heptapod.net/pypy/pypy/-/commit/77bb90aa67ec2b76748c9b96992fc3c3c4bf3fb5) to PyPy (for all python versions 2.7, 3.6, 3.7)

**Comment by YannickJadoul** [comment]
Oh, even a patch to PyPy? That might also work!
I was thinking of a patch to the generated `requirements.txt`/`constraints.txt`, building on @henryiii's suggestion. But maybe a temporary patch to PyPy might be even nicer, yes.

**Comment by Czaki** [comment]
It works. :)

**Comment by joerick** [comment]
Fixed in #502

**Comment by joerick** [comment]
Released as 1.7.3
